### ISSUE TITLE
Integrate Deploy into Cocoapods

### ISSUE THREAD

**Issue by jcampbell05** [CocoaPods/CocoaPods#6943]
# Report

A while back I built https://github.com/jcampbell05/cocoapods-deploy/ as a quick way of implementing the `bundle deploy` functionality. Sometimes all you want is to simply install the source for the pods without having to incur the cost of checking versions or the local pod repository on the machine.

I also want the security of not installing the incorrect version if for example the lockfile doesn't yet specify a pod version and you run `pod install`.

Back in the days of Cocoapods pre-beta and 1.0 I was able to achieve a speed up of 73% which had a huge effect when building projects under Travis CI.

Whilst these speedups may not be as large now the amazing work dantomi did with 1.3.0 :) I still think there is a need for a command like this.

For me as the plugin's author I'm finding that it relies on a few ugly monkey patches and is quite tied into how Cocoapod works so I believe it should belong in the core.

However even if you disagree or even if this never happens I think there are a few tweaks I would like to contribute.

# How does it work ?

We check for and read the podfile as well as the lockfile. We figure out the location to download the source code from for each pod based on the source of the podfile and download it for the specified version in the lockfile.

Before we trigger the installation step, we transform the dependencies specified in the podfile, so that they all point to the downloaded source code. Without this the install step will throw an error since the `deploy` command stops the local repositories on the machine from being updated.

Once this is done, the user should be able to build their project as normal

# What do I need to do ?

Below I'll outline the tweaks I need to make this feasible or even just to help improve the cocoapods core.

This will be my first time contributing so would appreciate help in implementing this.

- [ ]  Add the `pod deploy` command specified here -  https://github.com/jcampbell05/cocoapods-deploy/blob/master/lib/cocoapods-deploy/command/deploy.rb
- [ ] Add the `DeployTransformer` class which transforms the podfile specifications to point at the downloaded copies of the source code for each pod - https://github.com/jcampbell05/cocoapods-deploy/blob/master/lib/cocoapods-deploy/deploy_transformer.rb
- [ ] Add / Refactor ? the `DeployDownloader` class which figure out how to download the source for a pod, we use our own class to turn off validation and to work with the `DeployTransformer` to figure out where to download them from (The original one needs the pod repo on the local computer- https://github.com/jcampbell05/cocoapods-deploy/blob/master/lib/cocoapods-deploy/deploy_downloader.rb
- [ ] We patch the resolve to change how dependency resolution is preformed, in how patched version we just make sure each resolved dependency has the correct version download - https://github.com/jcampbell05/cocoapods-deploy/blob/master/lib/cocoapods-deploy/command/deploy.rb#L48
- [ ] We patch the ExternalSources to allow us to disable validation for certain podspecs, without this not all pods can be installed (see comments) https://github.com/jcampbell05/cocoapods-deploy/blob/master/lib/cocoapods-deploy/command/deploy.rb#L91
- [ ] We have to subclass the installer to be able to use our own analyse, to disable pre-downloads and disable modification to the lockfile - https://github.com/jcampbell05/cocoapods-deploy/blob/master/lib/cocoapods-deploy/deploy_installer.rb#L5
- [ ] We have to subclass the analyser to stop it returning repo sources that cocoapods normall tries to install (even when we disable installing the podspec repos with the standard flag) - https://github.com/jcampbell05/cocoapods-deploy/blob/master/lib/cocoapods-deploy/deploy_analyzer.rb

Would love to know your thoughts on what you think I could contribute :)

**Comment by orta** [comment]
This sounds like a big feature that would have an impact on a lot of things, are you willing to stick around to build this out and maintain it? I'd rather not include no-essential anymore giving how scarce the team is on time.

**Comment by jcampbell05** [comment]
@orta Yes I would be very willing to add this and I think this would be useful for a lot of projects I use it with.

I could start by just by allowing the plugin to work with CP without monkey patches.

**Comment by dnkoutso** [comment]
@jcampbell05 still any appetite for this? Otherwise I think I am going to close it.

**Comment by jcampbell05** [comment]
@dnkoutso I'm still working on removing the monkey patches should have a couple of PRs soon
